### ISSUE TITLE
Several Issues

### ISSUE THREAD

**Issue by chaoticmachinery** [blackbagtech/sleuthkit-APFS#7]
First off, I've been doing using the tools under a live image. I've seen some issues with fls and pstat.

fls isn't parsing the entire file system. Below is the fls command I am using:

fls  -m "" -p -r  -o 409640 -B 775896  /dev/rdisk0 

On my file system I have a known_hosts file under /Users/K/.ssh, which does not show up in the fls output. The output does list the /Users/K/.ssh directory, but nothing underneath it. I am sure there are other examples, that is the one I would identify easily. if it helps here is the output from the fls command for the .ssh directory.

/Users/K/.ssh|1095177|d/----------|0|0|0|0|0|0|0

The second issue I am seeing is related to fls. Running the same command as shown above at different times (say 5 to 10 min. a part) does not produce the same file. Grant I do realize that given on a live system, files are changing. So there should be some differences. However, not a 41 meg difference. Out2 has more files in it than out. 
-rw-r--r--    1 root root 130M Mar  5 15:45 out
-rw-r--r--    1 root root 171M Mar  5 15:45 out2

$ wc -l out*
   750806 out
   938075 out2
  1688881 total

The third issue I've see with pstat is that I have observed the APSB block number changing through out the day. Below is the pstat command I am using:
pstat -o 409640 /dev/rdisk0 | grep "APSB"

Below is the output from the command"
 pstat -o 409640 /dev/rdisk0 | grep "APSB"
|   APSB Block Number: 778844
|   APSB oid: 1027
|   APSB xid: 61234

After waiting 20 min. I get the following output:
  pstat -o 409640 /dev/rdisk0 | grep "APSB"
|   APSB Block Number: 779294
|   APSB oid: 1027
|   APSB xid: 61236

My guess is that with the APSB changing it is affecting the output from the fls command. Hence either why files are missing and other inconsistencies.

**Comment by simsong** [comment]
If you unmount the live image, do you see the `known_hosts` ? If so, the blocks may not be committed.

**Comment by jtsylve** [comment]
You'd be surprised how much changes very quickly on disk with APFS.  The copy-on-write nature of the file system means that even if you change a single bit of data, then not only will 4 KiB of data need to be rewritten elsewhere, but all blocks referencing the changed block will need to be rewritten, (and the blocks referencing those blocks and so on)...

Due to this, APFS metadata objects aren't always immediately flushed to disk.  In addition both the TSK image layer and this APFS layer contain caches, so reading from the same block twice expecting different results will produce undesired results.  Running these commands on a live partition that is mounted read/write will often fail in unpredictable ways because of this.  The APSB (or blocks pointed to it down the line) can even become invalid in the small time between running pstat and fls.

TL;DR - Unless you're seeing this behavior on an unmounted image, what you're seeing is not a bug in the implementation.  It's just a limitation of TSK when run against constantly changing data.

**Comment by chaoticmachinery** [comment]
Joe and Simson,

Thank you for getting back to me. Your explanations makes sense.

Thank your for your time,
K Murphy

Quoting Joe Sylve <notifications@github.com>:
